src/creatures/players/cast/cast_viewer.cpp

```diff
@@ -0,0 +1,1550 @@
+	for (const auto &[client, _] : m_viewers) {
+		client->sendSessionEndInformation(SESSION_END_LOGOUT);
+	}
+
+	m_viewers.clear();
+	m_mutes.clear();
+	removeCaster();
+
+	m_id = 0;
+	if (!full) {
+		return;
+	}
+
+	m_bans.clear();
+	m_castPassword = "";
+	m_castBroadcast = false;
+	m_castBroadcastTime = 0;
+	m_castLiveRecord = 0;
+}
+
+bool CastViewer::checkPassword(const std::string &_password) {
+	if (m_castPassword.empty()) {
+		return true;
+	}
+
+	std::string t = _password;
+	trimString(t);
+
+	return t == m_castPassword;
+}
+
+void CastViewer::setKickViewer(std::vector<std::string> list) {
+	for (const auto &it : list) {
+		for (const auto &[client, clientInfo] : m_viewers) {
+			if (asLowerCaseString(clientInfo.first) == it) {
+				client->sendSessionEndInformation(SESSION_END_LOGOUT);
+			}
+		}
+	}
+}
+
+const std::vector<std::string> &CastViewer::getMuteCastList() const {
+	return m_mutes;
+}
+
+void CastViewer::setMuteViewer(std::vector<std::string> mutes) {
+	m_mutes = mutes;
+}
+
+const std::map<std::string, uint32_t> &CastViewer::getBanCastList() const {
+	return m_bans;
+}
+
+void CastViewer::setBanViewer(std::vector<std::string> bans) {
+	std::vector<std::string>::const_iterator it;
+	for (auto bit = m_bans.begin(); bit != m_bans.end();) {
+		it = std::find(bans.begin(), bans.end(), bit->first);
+		if (it == bans.end()) {
+			m_bans.erase(bit++);
+		} else {
+			++bit;
+		}
+	}
+
+	for (it = bans.begin(); it != bans.end(); ++it) {
+		for (const auto &[client, clientInfo] : m_viewers) {
+			if (asLowerCaseString(clientInfo.first) != *it) {
+				continue;
+			}
+
+			m_bans[*it] = client->getIP();
+			client->sendSessionEndInformation(SESSION_END_LOGOUT);
+		}
+	}
+}
+
+bool CastViewer::checkBannedIP(uint32_t ip) const {
+	for (const auto &it : m_bans) {
+		if (it.second == ip) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+std::shared_ptr<ProtocolGame> CastViewer::getCastOwner() const {
+	return m_owner;
+}
+
+void CastViewer::setCastOwner(std::shared_ptr<ProtocolGame> client) {
+	m_owner = client;
+}
+
+void CastViewer::resetCastOwner() {
+	m_owner.reset();
+}
+
+std::string CastViewer::getCastPassword() const {
+	return m_castPassword;
+}
+
+void CastViewer::setCastPassword(const std::string &value) {
+	m_castPassword = value;
+}
+
+bool CastViewer::isCastBroadcasting() const {
+	return m_castBroadcast;
+}
+
+void CastViewer::setCastBroadcast(bool value) {
+	m_castBroadcast = value;
+}
+
+std::string CastViewer::getCastBroadcastTimeString() const {
+	std::stringstream broadcast_message;
+	int64_t seconds = getCastBroadcastTime() / 1000;
+	uint16_t hour = floor(seconds / 60 / 60 % 24);
+	uint16_t minute = floor(seconds / 60 % 60);
+	uint16_t second = floor(seconds % 60);
+
+	if (hour > 0) {
+		broadcast_message << hour << " hours, ";
+	}
+	if (minute > 0) {
+		broadcast_message << minute << " minutes and ";
+	}
+	broadcast_message << second << " seconds.";
+	return broadcast_message.str();
+}
```

Comment:
I think that it could be simplier showing the hours and minutes even if they are lesser than 1;
```suggestion
std::string CastViewer::getCastBroadcastTimeString() const {
	int64_t seconds = getCastBroadcastTime() / 1000;
	uint16_t hour = floor(seconds / 60 / 60 % 24);
	uint16_t minute = floor(seconds / 60 % 60);
	uint16_t second = floor(seconds % 60);

	return fmt::format("{} hours, {} minutes and {} seconds", hour, minute, second);
}
```